File: src/lib/api/archive.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  ArchiveHealth,
  ArchiveInitializationResult,
  ArchiveLocalSnapshotSummary,
  ArchiveSearchFilters,
  ArchiveSearchPage,
  ArchiveSettings,
  ArchivedConversationDetail,
  HistoryImportPreview,
  HistoryImportResult,
  LocalConversationApiStatus,
} from "@/types";

export const archiveApi = {
  async initialize(
    archive: ArchiveSettings,
  ): Promise<ArchiveInitializationResult> {
    return await invoke("initialize_conversation_archive", { archive });
  },

  async health(): Promise<ArchiveHealth> {
    return await invoke("get_archive_health");
  },

  async initializeLocal(): Promise<boolean> {
    return await invoke("initialize_local_conversation_archive");
  },

  async triggerLocalHistoryImport(): Promise<boolean> {
    return await invoke("trigger_local_history_import");
  },

  async localApiStatus(): Promise<LocalConversationApiStatus> {
    return await invoke("get_local_conversation_api_status");
  },

  async revealLocalApiToken(): Promise<string> {
    return await invoke("reveal_local_conversation_api_token");
  },

  async rotateLocalApiToken(): Promise<string> {
    return await invoke("rotate_local_conversation_api_token");
  },

  async listLocalSnapshots(): Promise<ArchiveLocalSnapshotSummary[]> {
    return await invoke("list_archive_local_snapshots");
  },

  async createLocalSnapshot(): Promise<ArchiveLocalSnapshotSummary> {
    return await invoke("create_archive_local_snapshot");
  },

  async restoreLocalSnapshot(id: string): Promise<void> {
    await invoke("restore_archive_local_snapshot", { id });
  },

  async deleteLocalSnapshot(id: string): Promise<void> {
    await invoke("delete_archive_local_snapshot", { id });
  },

  async previewImport(): Promise<HistoryImportPreview> {
    return await invoke("preview_local_history_import");
  },

  async importLocalHistory(): Promise<HistoryImportResult> {
    return await invoke("import_local_history");
  },

  async search(options: {
    query: string;
    filters: ArchiveSearchFilters;
    cursor?: string;
    pageSize?: number;
  }): Promise<ArchiveSearchPage> {
    return await invoke("search_archived_conversations", {
      query: options.query,
      filters: options.filters,
      cursor: options.cursor ?? null,
      pageSize: options.pageSize ?? 50,
    });
  },

  async get(id: string): Promise<ArchivedConversationDetail> {
    return await invoke("get_archived_conversation", { id });
  },

  async export(
    ids: string[],
    format: "json" | "markdown",
    targetPath: string,
  ): Promise<boolean> {
    return await invoke("export_archived_conversations", {
      ids,
      format,
      targetPath,
    });
  },

  async delete(ids: string[]): Promise<{ deleted: number }> {
    return await invoke("delete_archived_conversations", { ids });
  },

  async testRedaction(input: string): Promise<string> {
    return await invoke("test_archive_redaction", { input });
  },
};

File: src-tauri/src/commands/archive.rs
#![allow(non_snake_case)]

use crate::archive::{
    ArchiveDeleteResult, ArchiveHealth, ArchiveInitializationResult, ArchiveLocalSnapshotSummary,
    ArchiveSearchFilters, ArchiveSearchPage, ArchivedConversationDetail, HistoryImportPreview,
    HistoryImportResult,
};
use crate::settings::ArchiveSettings;
use crate::store::AppState;

#[tauri::command]
pub async fn get_archive_health(
    state: tauri::State<'_, AppState>,
) -> Result<ArchiveHealth, String> {
    Ok(state.archive.health().await)
}

#[tauri::command]
pub async fn initialize_conversation_archive(
    state: tauri::State<'_, AppState>,
    archive: ArchiveSettings,
) -> Result<ArchiveInitializationResult, String> {
    state.archive.initialize(archive).await
}

#[tauri::command]
pub async fn initialize_local_conversation_archive(
    state: tauri::State<'_, AppState>,
) -> Result<bool, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || archive.prepare_local_archive())
        .await
        .map_err(|e| format!("本机归档初始化任务失败: {e}"))??;
    Ok(true)
}

#[tauri::command]
pub async fn trigger_local_history_import(
    state: tauri::State<'_, AppState>,
) -> Result<bool, String> {
    Ok(state.archive.trigger_local_history_import())
}

#[tauri::command]
pub async fn get_local_conversation_api_status(
    state: tauri::State<'_, AppState>,
) -> Result<crate::archive::local_api::LocalConversationApiStatus, String> {
    Ok(crate::archive::local_api::status(&state.archive))
}

#[tauri::command]
pub async fn reveal_local_conversation_api_token() -> Result<String, String> {
    crate::archive::local_api::get_or_create_token()
}

#[tauri::command]
pub async fn rotate_local_conversation_api_token() -> Result<String, String> {
    crate::archive::local_api::rotate_token()
}

#[tauri::command]
pub async fn preview_local_history_import(
    state: tauri::State<'_, AppState>,
) -> Result<HistoryImportPreview, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || archive.preview_local_history())
        .await
        .map_err(|e| format!("本机历史扫描任务失败: {e}"))?
}

#[tauri::command]
pub async fn import_local_history(
    state: tauri::State<'_, AppState>,
) -> Result<HistoryImportResult, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || archive.import_local_history())
        .await
        .map_err(|e| format!("本机历史导入任务失败: {e}"))?
}

#[tauri::command]
pub async fn search_archived_conversations(
    state: tauri::State<'_, AppState>,
    query: String,
    filters: ArchiveSearchFilters,
    cursor: Option<String>,
    pageSize: usize,
) -> Result<ArchiveSearchPage, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || {
        archive.search(&query, &filters, cursor.as_deref(), pageSize)
    })
    .await
    .map_err(|e| format!("归档检索任务失败: {e}"))?
}

#[tauri::command]
pub async fn get_archived_conversation(
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<ArchivedConversationDetail, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || archive.detail(&id))
        .await
        .map_err(|e| format!("读取归档会话任务失败: {e}"))?
}

#[tauri::command]
pub async fn export_archived_conversations(
    state: tauri::State<'_, AppState>,
    ids: Vec<String>,
    format: String,
    targetPath: String,
) -> Result<bool, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || {
        archive.export(&ids, &format, std::path::Path::new(&targetPath))
    })
    .await
    .map_err(|e| format!("归档导出任务失败: {e}"))??;
    Ok(true)
}

#[tauri::command]
pub async fn delete_archived_conversations(
    state: tauri::State<'_, AppState>,
    ids: Vec<String>,
) -> Result<ArchiveDeleteResult, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || archive.delete(&ids))
        .await
        .map_err(|e| format!("归档删除任务失败: {e}"))?
}

#[tauri::command]
pub async fn list_archive_local_snapshots(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<ArchiveLocalSnapshotSummary>, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || archive.list_local_snapshots())
        .await
        .map_err(|e| format!("读取归档本地快照任务失败: {e}"))?
}

#[tauri::command]
pub async fn create_archive_local_snapshot(
    state: tauri::State<'_, AppState>,
) -> Result<ArchiveLocalSnapshotSummary, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || archive.create_local_snapshot_now())
        .await
        .map_err(|e| format!("创建归档本地快照任务失败: {e}"))?
}

#[tauri::command]
pub async fn restore_archive_local_snapshot(
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<ArchiveLocalSnapshotSummary, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || archive.restore_local_snapshot(&id))
        .await
        .map_err(|e| format!("恢复归档本地快照任务失败: {e}"))?
}

#[tauri::command]
pub async fn delete_archive_local_snapshot(
    state: tauri::State<'_, AppState>,
    id: String,
) -> Result<bool, String> {
    let archive = state.archive.clone();
    tauri::async_runtime::spawn_blocking(move || archive.delete_local_snapshot(&id))
        .await
        .map_err(|e| format!("删除归档本地快照任务失败: {e}"))??;
    Ok(true)
}

#[tauri::command]
pub async fn test_archive_redaction(
    state: tauri::State<'_, AppState>,
    input: String,
) -> Result<String, String> {
    state.archive.test_redaction(&input)
}
